Homework 16.

IgushkinV opened the pull request:
Рефакторинг классов из 15 домашки.
Реализация интерфейса IRepository.
Без 5 и 6. Без тестов.

IgushkinV committed:
Homework 16.
Рефакторинг классов из 15 домашки.
Без 5 и 6. Без тестов.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/Main.java:
```diff
@@ -1,77 +1,53 @@
-        orderUtils.delete(3);
-    }
+        ApplicationContext applicationContext = new AnnotationConfigApplicationContext(ApplicationConfig.class);
+        var customerRepository = (CustomerRepository) applicationContext.getBean("customerRepository");
```
Лучше демонстрировать работу сервисов.

IgushkinV replied:
Пока не переделал. Сервисы сделал, но они принимают jsonObject из сервлетов на вход.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,42 @@
+
+@Configuration
+@ComponentScan("homework-impl/src/main/java/com/epam/igushkin/homework")
+public class ApplicationConfig {
```
Файлы конфигурации создают под каждую необходимую группу. Есть классы относящиеся к Jpa (entityManagerFactory) то это JpaConfig, есть классы отосящиеся к настройке локализации и работы с ресурсами (MessageSource), то это I18nConfig и т.д.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/repository/IRepository.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.Optional;
+
+public interface IRepository<T> {
+    public T create(T entity);
```
JavaDoc

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/repository/IRepository.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface IRepository<T> {
```
Давай придерживаться более популярного наименования в Java. Где интерфесы просто *Repository, а реализации *RepositoryImpl

IgushkinV replied:
Переделал.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/repository/IRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+    public List<T> readAll();
+
+    public boolean update(T entity);
```
Лучше возвращать объект

IgushkinV replied:
Переделал.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Bean
+    public CustomerRepository customerRepository() {
+        return new CustomerRepository(entityManagerFactory());
```
Репозитории можно сделать просто @ Component и автовайрить туда фактори

IgushkinV replied:
Получается, на весь конфиг будет только EntityManagerFactory? Если остальные бины будут автовариться.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,42 @@
+import javax.persistence.Persistence;
+
+@Configuration
+@ComponentScan("homework-impl/src/main/java/com/epam/igushkin/homework")
```
Когда ты скомпилишь приложения у тебя не будет в пути "homework-impl/src/main/java"

IgushkinV replied:
Да, что-то не то выбрал в CopyPath.

tlt-java-courses commented on homework-impl/src/main/java/com/epam/igushkin/homework/servlets/CustomerServlet.java:
```diff
@@ -1,37 +1,42 @@
 public class CustomerServlet extends HttpServlet {
 
-    private final CustomerUtils customerUtils = new CustomerUtils();
+    private final IRepository<Customer> customerRepository;
```
API слой ничего не должен знать о персистентном слое

IgushkinV replied:
Но должен общаться с сервисным, правильно?

IgushkinV replied:
Переделал.

tlt-java-courses reviewed (changes requested):
Где сервисный слой с бизнес логикой?
Тесты

IgushkinV committed:
Homework 16.
Создан сервисный слой, который через аннотацию CRUDService принимает данные из сервлетов и отправляет запросы к слою репозиториев.

Исправлены небольшие замечания.

tlt-java-courses commented:
1. Не выполняется update скриптов миграции
![image](https://user-images.githubusercontent.com/75726498/120905457-3220bb80-c663-11eb-8d9e-28dc6e238b7a.png)
2. Ошибка в приложении при выполнении
![image](https://user-images.githubusercontent.com/75726498/120905485-609e9680-c663-11eb-83c0-9d5e0b47a2d1.png)
3. Репозитории не в качестве компонентов, а бинами в конфигурации
![image](https://user-images.githubusercontent.com/75726498/120905506-7ad87480-c663-11eb-8280-4e5c204a81a0.png)
4. Такого пути для компонент скана не существует
![image](https://user-images.githubusercontent.com/75726498/120905517-98a5d980-c663-11eb-9ae7-15d10fdc2771.png)
5. Не выполнение задания со стабом данных без БД
6. Сервисы относятся к слою с бизнес логикой. CRUD - терминология персистентного слоя. (смешение логики)
![image](https://user-images.githubusercontent.com/75726498/120905547-d86cc100-c663-11eb-831c-fd66d9da1338.png)
7. Внутри системы мы должны оперировать сущностями, а не приходящими извне данными, что обрабатываются на презентационном слое (АПИ) (смешение логики)
![image](https://user-images.githubusercontent.com/75726498/120905569-fcc89d80-c663-11eb-8fe3-f996314a9930.png)
8. В презнтационном слое работа с внутренними сущностями системы (смешение логики)
![image](https://user-images.githubusercontent.com/75726498/120905589-22ee3d80-c664-11eb-93d5-48453030cf7c.png)
9. Нет тестов

10. Не соблюдение клин кода
10.1. Лишний класс 
![image](https://user-images.githubusercontent.com/75726498/120905510-8a57bd80-c663-11eb-8c03-ad50c8eeb926.png)
10.2. JavaDocs
